Checkpoint & backup scheduler impl

panagiotisgts opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

Introduce the Checkpoint Scheduler who is responsible for injecting checkpoints based on the configured intervals in the cluster.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #40442
closes #40443

panagiotisgts committed:
fix: confirm checkpoint with correct type

panagiotisgts committed:
feat: provision checkpoint type when requesting a backup internally

panagiotisgts committed:
fix: parse blank schedules as none

panagiotisgts committed:
feat: introduce checkpoint scheduler

Brokers now register a checkpoint scheduling service during startup if continuous backups are enabled.

A broker is eligible to handle checkpoint creation if it's the lowest member id in the cluster based on SWIM membership.

panagiotisgts committed:
test: coverage for checkpoint scheduler startup and behavior

panagiotisgts committed:
fix: supply checkpoint type for old tests

panagiotisgts committed:
refactor: replace tuples with records and clean up future chain for clarity

panagiotisgts committed:
refactor: replace actor.call with run to fix failing controlled actor test

lenaschoenburg commented on zeebe/backup/src/main/java/io/camunda/zeebe/backup/schedule/CheckpointScheduler.java:
```diff
@@ -0,0 +1,261 @@
+        .thenApply(this::peekDelay)
+        .thenApply(Duration::ofMillis)
+        .exceptionally(this::onError)
+        .thenAccept(delay -> run(() -> schedule(delay, this::reschedulingTask)));
```
❌ This is bypassing the actor almost completely which is defeating its purpose but also incorrect because methods like `onError` modify internal state which should only be done within an actor job.
As discussed, please wrap `CompletableFuture`s received from the `backupRequestHandler` in a `CompletableActorFuture` and use that to set up the execution chain.

panagiotisgts committed:
refactor: remove isStopped flag

panagiotisgts committed:
refactor: extract decision functions as providers

panagiotisgts committed:
fix: correctly utilize actor

Replace CompletableFutures with Actor ones, to utilize the submitted actor. Also, limits the state modifications to the Actor thread alone.

panagiotisgts committed:
docs: document checkpoint threshold

panagiotisgts committed:
test: fix scheduler stop test case

lenaschoenburg commented on zeebe/backup/src/main/java/io/camunda/zeebe/backup/schedule/CheckpointScheduler.java:
```diff
@@ -0,0 +1,261 @@
+      checkpointDecider = this::nextBackup;
+      executorDelayProvider = this::backupScheduleDelayProvider;
+    }
+  }
```
@panagiotisgts With fcb4d2e this didn't really change, we are still picking a provider during construction. I think I still would prefer to not have these providers at all, just call into one method which checks the current config and determines the delay/next schedule action on the fly.

lenaschoenburg commented on zeebe/backup/src/main/java/io/camunda/zeebe/backup/schedule/CheckpointScheduler.java:
```diff
@@ -47,77 +51,122 @@ public CheckpointScheduler(
+        .min(Comparator.comparingLong(PartitionCheckpointState::checkpointId))
+        .map(PartitionCheckpointState::checkpointTimestamp)
+        .map(Instant::ofEpochMilli)
+        .orElseGet(() -> schedule == null ? null : previousExecution(schedule, now));
```
❓ Why do we fall back to `previousExecution` here? We received no `PartitionCheckpointState`, but then just assume that the previous execution was successful? Wouldn't it be safer to assume it _wasn't_ successful and then schedule the next execution as soon as possible?

panagiotisgts replied:
The previous execution is calculated based on the schedules definition. So when calculating the previous, you essentially will trigger the next invocation instantly, since you go back the exact interval of each schedule.

You don't assume something here. You just avoid having to deal with different interpretations of what _previous_ means when no state is present, so you create an fake state :smile:

panagiotisgts committed:
refactor: remove decider functions, move delay calculation in the onComplete step

panagiotisgts committed:
fix: wrap state acquisition within actor

lenaschoenburg reviewed (approved):
The implementation is in great shape, kudos 🚀 

🔧 I'd like to work on the tests, they are heavily relying on complicated mock setups and feel very brittle. I don't have any concrete suggestions on how to tackle this so I don't want to prevent this from merging. Perhaps we can schedule something to go over it together and find a different way to test this?

panagiotisgts committed:
fix: self correct schedule on potential drift

A schedules determined execution can be in the past if the previous was too far behind, that results in an instant checkpoint. However, if a schedule is too far behind the next execution could still be in the past resulting in longer than expected delays since we are taking the absolute value of the delay.

With this change we always correct the instructions checkpoint time upon completion, so a drifted schedule will self correct.
